src/modules/kvs/commit.c

```diff
@@ -369,6 +376,116 @@ static int commit_unroll (commit_t *c, int current_epoch, json_t *dir)
+                                     json_t *val, href_t ref)
+{
+    struct cache_entry *hp;
+    json_t *val_data;
+    int ret;
+
+    if (!(val_data = treeobj_get_data (val)))
+        return -1;
+
+    if ((ret = store_cache (c, current_epoch, val_data,
+                            true, ref, &hp)) < 0)
+        return -1;
+
+    if (ret) {
+        if (zlist_push (c->dirty_cache_entries_list, hp) < 0) {
+            commit_cleanup_dirty_cache_entry (c, hp);
+            errno = ENOMEM;
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int commit_append (commit_t *c, int current_epoch, json_t *dirent,
+                          json_t *dir, const char *final_name)
+{
+    json_t *entry;
+
+    if (!treeobj_is_val (dirent)) {
+        errno = EPROTO;
+        return -1;
+    }
+
+    entry = treeobj_get_entry (dir, final_name);
+
+    if (!entry) {
+        /* entry not found, treat like normal insertion */
+        if (treeobj_insert_entry (dir, final_name, dirent) < 0)
+            return -1;
+    }
+    else if (treeobj_is_valref (entry)) {
+        href_t ref;
+
+        /* treeobj is valref, so we need to append the new data's
+         * blobref to this tree object.  Before doing so, we must save
+         * off the new data to the cache and mark it dirty for
+         * flushing later (if necessary) */
+
+        if (commit_val_data_to_cache (c, current_epoch,
+                                      dirent, ref) < 0)
+            return -1;
+
+        if (treeobj_append_blobref (entry, ref) < 0)
+            return -1;
+    }
+    else if (treeobj_is_val (entry)) {
+        json_t *tmp;
+        href_t ref1, ref2;
+
+        /* treeobj entry is val, so we need to convert the treeobj
+         * into a valref first.  Then the procedure is basically the
+         * same as the treeobj valref case above.
+         */
+
+        if (commit_val_data_to_cache (c, current_epoch,
+                                      entry, ref1) < 0)
+            return -1;
+
+        if (commit_val_data_to_cache (c, current_epoch,
+                                      dirent, ref2) < 0)
+            return -1;
+
+        if (!(tmp = treeobj_create_valref (ref1)))
+            return -1;
+
+        if (treeobj_append_blobref (tmp, ref2) < 0) {
+            json_decref (tmp);
+            return -1;
+        }
+
+        if (treeobj_insert_entry (dir, final_name, tmp) < 0) {
+            json_decref (tmp);
+            return -1;
+        }
+    }
+    else if (treeobj_is_symlink (entry)) {
+        /* Could use EPERM - operation not permitted, but want to
+         * avoid confusion with "common" errnos, we'll use this one
+         * instead. */
+        errno = EOPNOTSUPP;
+        return -1;
+    }
+    else if (treeobj_is_dir (entry)
+             || treeobj_is_dirref (entry)) {
+        errno = EISDIR;
+        return -1;
+    }
+    else {
+        char *s = json_dumps (entry, 0);
```

Comment:
use treeobj_encode()?

Reply:
Since this is the "corrupt treeobj" path, I think using json_dumps() is the right idea.  

Using treeobj_encode() works right now b/c it presently just calls json_dumps() directly and does nothing else.  Could treeobj_encode() ever check for treeobj_validate?